### PR TITLE
Rollout migrate table to grid

### PR DESCRIPTION
**Pull request by asharani-murugesh**
Rollout migrate table to grid 

Signed-off-by: asharani-murugesh <asharani.murugesh@in.bosch.com>

### CODE REVIEW

File: hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/customrenderers/client/renderers/HtmlLabelRenderer.java [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+
+import java.util.HashMap;
+import java.util.Map;
+
+import com.google.gwt.core.client.GWT;
+import com.vaadin.client.renderers.WidgetRenderer;
+import com.vaadin.client.ui.VLabel;
+import com.vaadin.client.widget.grid.RendererCellReference;
+
+public class HtmlLabelRenderer extends WidgetRenderer<String, VLabel> {
```
Comment (SirWayne) [inline hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/customrenderers/client/renderers/HtmlLabelRenderer.java]: Add javadoc

File: hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/customrenderers/client/renderers/HtmlButtonRenderer.java [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.eclipse.hawkbit.ui.customrenderers.client.renderers;
+
+import com.google.gwt.user.client.ui.Button;
+import com.vaadin.client.renderers.ButtonRenderer;
+import com.vaadin.client.ui.VButton;
+import com.vaadin.client.widget.grid.RendererCellReference;
+
+public class HtmlButtonRenderer extends ButtonRenderer {
```
Comment (SirWayne) [inline hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/customrenderers/client/renderers/HtmlButtonRenderer.java]: add javadoc

File: hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/customrenderers/client/LinkRendererConnector.java [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.eclipse.hawkbit.ui.customrenderers.client;
+
+import com.vaadin.client.connectors.ButtonRendererConnector;
+import com.vaadin.shared.ui.Connect;
+
+@Connect(org.eclipse.hawkbit.ui.customrenderers.renderers.LinkRenderer.class)
+public class LinkRendererConnector extends ButtonRendererConnector {
```
Comment (SirWayne) [inline hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/customrenderers/client/LinkRendererConnector.java]: add javadoc

File: hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/customrenderers/client/HtmlLabelRendererConnector.java [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.eclipse.hawkbit.ui.customrenderers.client;
+
+import com.vaadin.client.connectors.AbstractRendererConnector;
+import com.vaadin.shared.ui.Connect;
+
+@Connect(org.eclipse.hawkbit.ui.customrenderers.renderers.HtmlLabelRenderer.class)
+public class HtmlLabelRendererConnector extends AbstractRendererConnector<String> {
```
Comment (SirWayne) [inline hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/customrenderers/client/HtmlLabelRendererConnector.java]: add javadoc

File: hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/customrenderers/client/HtmlButtonRendererConnector.java [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.eclipse.hawkbit.ui.customrenderers.client;
+
+import com.vaadin.client.connectors.ButtonRendererConnector;
+import com.vaadin.shared.ui.Connect;
+
+@Connect(org.eclipse.hawkbit.ui.customrenderers.renderers.HtmlButtonRenderer.class)
+public class HtmlButtonRendererConnector extends ButtonRendererConnector {
```
Comment (SirWayne) [inline hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/customrenderers/client/HtmlButtonRendererConnector.java]: add javadoc

File: hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/customrenderers/renderers/LinkRenderer.java [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.eclipse.hawkbit.ui.customrenderers.renderers;
+
+import com.vaadin.ui.renderers.ButtonRenderer;
+
+public class LinkRenderer extends ButtonRenderer {
```
Comment (SirWayne) [inline hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/customrenderers/renderers/LinkRenderer.java]: Add javadoc

File: hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/customrenderers/renderers/HtmlLabelRenderer.java [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.eclipse.hawkbit.ui.customrenderers.renderers;
+
+import com.vaadin.ui.Grid.AbstractRenderer;
+
+public class HtmlLabelRenderer extends  AbstractRenderer<String> {
```
Comment (SirWayne) [inline hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/customrenderers/renderers/HtmlLabelRenderer.java]: add javadoc

File: hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/customrenderers/client/renderers/LinkRenderer.java [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.eclipse.hawkbit.ui.customrenderers.client.renderers;
+
+import com.google.gwt.user.client.ui.Button;
+import com.vaadin.client.renderers.ButtonRenderer;
+import com.vaadin.client.ui.VButton;
+import com.vaadin.client.widget.grid.RendererCellReference;
+
+public class LinkRenderer extends ButtonRenderer {
```
Comment (SirWayne) [inline hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/customrenderers/client/renderers/LinkRenderer.java]: add javadoc

File: hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/customrenderers/renderers/HtmlButtonRenderer.java [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.eclipse.hawkbit.ui.customrenderers.renderers;
+
+import com.vaadin.ui.renderers.ButtonRenderer;
+
+public class HtmlButtonRenderer extends ButtonRenderer {
```
Comment (SirWayne) [inline hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/customrenderers/renderers/HtmlButtonRenderer.java]: add javadoc

File: hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/customrenderers/renderers/HtmlButtonRenderer.java [hunk 10]
```diff
@@ -0,0 +1,22 @@
...
+    private static final long serialVersionUID = -1242995370043404892L;
+    public HtmlButtonRenderer() {
+        super();
+    }
+    public HtmlButtonRenderer(RendererClickListener listener) {
+        super(listener);
+    }
+    
+}
```
Comment (SirWayne) [inline hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/customrenderers/renderers/HtmlButtonRenderer.java]: add empty line

File: hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/customrenderers/renderers/LinkRenderer.java [hunk 11]
```diff
@@ -0,0 +1,22 @@
...
+    private static final long serialVersionUID = -1242995370043404892L;
+    public LinkRenderer() {
+        super();
+    }
+    public LinkRenderer(RendererClickListener listener) {
+        super(listener);
+    }
+    
+}
```
Comment (SirWayne) [inline hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/customrenderers/renderers/LinkRenderer.java]: add empty line

File: hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/rollout/DistributionBarHelper.java [hunk 12]
```diff
@@ -0,0 +1,134 @@
...
+        // + getPartWidth(value, totalValue, noOfParts) + ";\" title = \"" +
+        // partValue + "\"><span class=\""
+        // + DISTRIBUTION_BAR_PART_VALUE_CLASSNAME + "\">" + value +
+        // "</span></div>";
+        return "<div class=\"" + getPartStyle(partIndex, noOfParts, partValue) + "\" style=\"width: "
+                + getPartWidth(value, totalValue, noOfParts) + ";\"><span class=\""
+                + DISTRIBUTION_BAR_PART_VALUE_CLASSNAME + "\">" + value + "</span></div>";
+    }
+
+    public static String getDistributionBarAsHTMLString(Map<Status, Long> statusTotalCountMap) {
```
Comment (SirWayne) [inline hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/rollout/DistributionBarHelper.java]: add javadoc

File: hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/rollout/DistributionBarHelper.java [hunk 13]
```diff
@@ -0,0 +1,134 @@
...
+    private static String getPartWidth(Long value, Long totalValue, int noOfParts) {
+        final double minTotalSize = MINIMUM_PART_SIZE * noOfParts;
+        final double availableSize = PARENT_SIZE_IN_PCT - minTotalSize;
+        double val = MINIMUM_PART_SIZE + (double) value / totalValue * availableSize;
+        return String.format("%.3f", val) + "%";
+    }
+
+    private static String getPart(int partIndex, Status status, Long value, Long totalValue, int noOfParts) {
+        String partValue = status.toString().toLowerCase();
+        // return "<div class=\"" + getPartStyle(partIndex, noOfParts,
```
Comment (SirWayne) [inline hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/rollout/DistributionBarHelper.java]: remove comment code

File: hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/rollout/DistributionBarHelper.java [hunk 14]
```diff
@@ -0,0 +1,134 @@
...
+            int partIndex = 1;
+            for (Map.Entry<Status, Long> entry : statusMapWithNonZeroValues.entrySet()) {
+                if (entry.getValue() > 0) {
+                    htmlString.append(getPart(partIndex, entry.getKey(), entry.getValue(), totalValue,
+                            statusMapWithNonZeroValues.size()));
+                    partIndex++;
+                }
+            }
+        } else {
+            return getUnintialisedBar();
```
Comment (SirWayne) [inline hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/rollout/DistributionBarHelper.java]: readability: return statement in the "true" block, then is no else necessary
 if (statusMapWithNonZeroValues.size() <= 0) {
  return getUnintialisedBar();
}
...

File: hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/rollout/DistributionBarHelper.java [hunk 15]
```diff
@@ -0,0 +1,134 @@
...
+        }
+        htmlString.append(getParentDivEnd());
+        return htmlString.toString();
+    }
+
+    public static Map<Status, Long> getNonZeroStatusList(Map<Status, Long> statusTotalCountMap) {
```
Comment (SirWayne) [inline hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/rollout/DistributionBarHelper.java]: add javadoc

File: hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/rollout/DistributionBarHelper.java [hunk 16]
```diff
@@ -0,0 +1,134 @@
...
+        return statusTotalCountMap.entrySet().stream().filter(p -> p.getValue() > 0)
+                .collect(Collectors.toMap(p -> p.getKey(), p -> p.getValue()));
+    }
+
+
+    public static String getTooltip(Map<Status, Long> statusCountMap) {
```
Comment (SirWayne) [inline hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/rollout/DistributionBarHelper.java]: add javadoc

File: hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/rollout/rollout/RolloutListGrid.java [hunk 17]
```diff
@@ -0,0 +1,568 @@
...
+    }
+
+    @PreDestroy
+    void destroy() {
+        eventBus.unsubscribe(this);
+    }
+
+    @EventBusListenerMethod(scope = EventScope.SESSION)
+    void onEvent(final RolloutEvent event) {
+        if (event == RolloutEvent.FILTER_BY_TEXT || event == RolloutEvent.CREATE_ROLLOUT
```
Comment (SirWayne) [inline hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/rollout/rollout/RolloutListGrid.java]: Use switch

File: hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/rollout/rollout/RolloutListGrid.java [hunk 18]
```diff
@@ -0,0 +1,568 @@
...
+    /**
+     * Handles the RolloutChangeEvent to refresh the item in the grid.
+     * 
+     * @param rolloutChangeEvent
+     *            the event which contains the rollout which has been changed
+     */
+    @SuppressWarnings("unchecked")
+    @EventBusListenerMethod(scope = EventScope.SESSION)
+    public void onEvent(final RolloutChangeEvent rolloutChangeEvent) {
+        if (rolloutUIState.isShowRollOuts()) {
```
Comment (SirWayne) [inline hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/rollout/rollout/RolloutListGrid.java]: early return for better readability:
 if (!rolloutUIState.isShowRollOuts()) {
return
}

File: hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/rollout/rollout/RolloutListGrid.java [hunk 19]
```diff
@@ -0,0 +1,568 @@
...
+            final Rollout rollout = rolloutManagement.findRolloutWithDetailedStatus(rolloutChangeEvent.getRolloutId());
+            final TotalTargetCountStatus totalTargetCountStatus = rollout.getTotalTargetCountStatus();
+            final LazyQueryContainer rolloutContainer = (LazyQueryContainer) getContainerDataSource();
+            final Item item = rolloutContainer.getItem(rolloutChangeEvent.getRolloutId());
+            if (null != item) {
```
Comment (SirWayne) [inline hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/rollout/rollout/RolloutListGrid.java]: early return

File: hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/rollout/rollout/RolloutListGrid.java [hunk 20]
```diff
@@ -0,0 +1,568 @@
...
+                item.getItemProperty(SPUILabelDefinitions.VAR_STATUS).setValue(rollout.getStatus());
+                item.getItemProperty(SPUILabelDefinitions.VAR_TOTAL_TARGETS_COUNT_STATUS)
+                        .setValue(totalTargetCountStatus);
+                final Long groupCount = (Long) item.getItemProperty(SPUILabelDefinitions.VAR_NUMBER_OF_GROUPS)
+                        .getValue();
+                if (null != rollout.getRolloutGroups() && groupCount != rollout.getRolloutGroups().size()) {
```
Comment (SirWayne) [inline hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/rollout/rollout/RolloutListGrid.java]: null not on the first paramter: use rollout.getRolloutGroups() != null

File: hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/rollout/rollout/RolloutListGrid.java [hunk 21]
```diff
@@ -0,0 +1,568 @@
...
+        contextMenu.open(event.getClientX(), event.getClientY());
+    }
+
+    private ContextMenu createContextMenu(final Long rolloutId) {
+        final Item row = getContainerDataSource().getItem(rolloutId);
+        final RolloutStatus rolloutStatus = (RolloutStatus) row.getItemProperty(SPUILabelDefinitions.VAR_STATUS)
+                .getValue();
+        final ContextMenu context = new ContextMenu();
+        context.addItemClickListener(event -> menuItemClicked(event));
+        if (rolloutStatus == RolloutStatus.READY) {
```
Comment (SirWayne) [inline hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/rollout/rollout/RolloutListGrid.java]: use switch

File: hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/rollout/rollout/RolloutListGrid.java [hunk 22]
```diff
@@ -0,0 +1,568 @@
...
+        }
+        if (permissionChecker.hasRolloutUpdatePermission()) {
+            final ContextMenuItem cancelItem = context.addItem(UPDATE_OPTION);
+            cancelItem.setData(new ContextMenuData(rolloutId, ACTION.UPDATE));
+        }
+        return context;
+    }
+
+    private String convertRolloutStatusToString(final RolloutStatus value) {
+        String result = null;
```
Comment (SirWayne) [inline hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/rollout/rollout/RolloutListGrid.java]: remove local paramter

File: hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/rollout/rollout/RolloutListGrid.java [hunk 23]
```diff
@@ -0,0 +1,568 @@
...
+        }
+        return result;
+    }
+
+    private void menuItemClicked(final ContextMenuItemClickEvent event) {
+        final ContextMenuItem item = (ContextMenuItem) event.getSource();
+        final ContextMenuData contextMenuData = (ContextMenuData) item.getData();
+        final Item row = getContainerDataSource().getItem(contextMenuData.getRolloutId());
+        final String rolloutName = (String) row.getItemProperty(SPUILabelDefinitions.VAR_NAME).getValue();
+        if (contextMenuData.getAction() == ACTION.PAUSE) {
```
Comment (SirWayne) [inline hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/rollout/rollout/RolloutListGrid.java]: switch

File: hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/rollout/rollout/RolloutListGrid.java [hunk 24]
```diff
@@ -0,0 +1,568 @@
...
+        if (SPUILabelDefinitions.VAR_STATUS.equals(cell.getPropertyId())) {
+            return cell.getProperty().getValue().toString().toLowerCase();
+        } else if (SPUILabelDefinitions.ACTION.equals(cell.getPropertyId())) {
+            return SPUILabelDefinitions.ACTION.toLowerCase();
+        } else if (SPUILabelDefinitions.VAR_NAME.equals(cell.getPropertyId())) {
+            return cell.getProperty().getValue().toString();
+        } else if (SPUILabelDefinitions.VAR_TOTAL_TARGETS_COUNT_STATUS.equals(cell.getPropertyId())) {
+            return DistributionBarHelper
+                    .getTooltip(((TotalTargetCountStatus) cell.getValue()).getStatusTotalCountMap());
+        } else {
```
Comment (SirWayne) [inline hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/rollout/rollout/RolloutListGrid.java]: else ist not necessary

File: hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/rollout/rolloutgroup/RolloutGroupListGrid.java [hunk 25]
```diff
@@ -0,0 +1,392 @@
...
+    }
+
+    @PreDestroy
+    void destroy() {
+        eventBus.unsubscribe(this);
+    }
+
+    @EventBusListenerMethod(scope = EventScope.SESSION)
+    void onEvent(final RolloutEvent event) {
+        if (event == RolloutEvent.SHOW_ROLLOUT_GROUPS) {
```
Comment (SirWayne) [inline hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/rollout/rolloutgroup/RolloutGroupListGrid.java]: 1. early return
2. use constant for first paramter

File: hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/rollout/rolloutgroup/RolloutGroupListGrid.java [hunk 26]
```diff
@@ -0,0 +1,392 @@
...
+import com.vaadin.data.util.converter.Converter;
+import com.vaadin.server.FontAwesome;
+import com.vaadin.spring.annotation.SpringComponent;
+import com.vaadin.spring.annotation.ViewScope;
+import com.vaadin.ui.renderers.ClickableRenderer.RendererClickEvent;
+import com.vaadin.ui.renderers.HtmlRenderer;
+
+@SpringComponent
+@ViewScope
+public class RolloutGroupListGrid extends AbstractGrid {
```
Comment (SirWayne) [inline hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/rollout/rolloutgroup/RolloutGroupListGrid.java]: add javadoc

File: hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/rollout/rolloutgroup/RolloutGroupListGrid.java [hunk 27]
```diff
@@ -0,0 +1,392 @@
...
+     * 
+     * 
+     * @param rolloutGroupChangeEvent
+     *            the event which contains the rollout group which has been
+     *            change
+     */
+    @SuppressWarnings("unchecked")
+    @EventBusListenerMethod(scope = EventScope.SESSION)
+    public void onEvent(final RolloutGroupChangeEvent rolloutGroupChangeEvent) {
+        if (rolloutUIState.isShowRolloutGroups()) {
```
Comment (SirWayne) [inline hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/rollout/rolloutgroup/RolloutGroupListGrid.java]: early return

File: hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/rollout/rolloutgroup/RolloutGroupListGrid.java [hunk 28]
```diff
@@ -0,0 +1,392 @@
...
+            final RolloutGroup rolloutGroup = rolloutGroupManagement
+                    .findRolloutGroupWithDetailedStatus(rolloutGroupChangeEvent.getRolloutGroupId());
+            final LazyQueryContainer rolloutContainer = (LazyQueryContainer) getContainerDataSource();
+            final Item item = rolloutContainer.getItem(rolloutGroup.getId());
+            if (item != null) {
```
Comment (SirWayne) [inline hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/rollout/rolloutgroup/RolloutGroupListGrid.java]: early return

File: hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/rollout/rolloutgroup/RolloutGroupListGrid.java [hunk 29]
```diff
@@ -0,0 +1,392 @@
...
+        columnsToBeHidden.add(SPUILabelDefinitions.VAR_DESC);
+        for (Object propertyId : columnsToBeHidden) {
+            getColumn(propertyId).setHidden(true);
+        }
+    }
+
+    @Override
+    protected CellDescriptionGenerator getDescriptionGenerator() {
+        return cell -> getDescription(cell);
+    };
```
Comment (SirWayne) [inline hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/rollout/rolloutgroup/RolloutGroupListGrid.java]: remove ;

File: hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/rollout/rolloutgroup/RolloutGroupListGrid.java [hunk 30]
```diff
@@ -0,0 +1,392 @@
...
+        if (SPUILabelDefinitions.VAR_STATUS.equals(cell.getPropertyId())) {
+            return cell.getProperty().getValue().toString().toLowerCase();
+        } else if (SPUILabelDefinitions.ACTION.equals(cell.getPropertyId())) {
+            return SPUILabelDefinitions.ACTION.toLowerCase();
+        } else if (SPUILabelDefinitions.VAR_NAME.equals(cell.getPropertyId())) {
+            return cell.getProperty().getValue().toString();
+        } else if (SPUILabelDefinitions.VAR_TOTAL_TARGETS_COUNT_STATUS.equals(cell.getPropertyId())) {
+            return DistributionBarHelper
+                    .getTooltip(((TotalTargetCountStatus) cell.getValue()).getStatusTotalCountMap());
+        } else {
```
Comment (SirWayne) [inline hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/rollout/rolloutgroup/RolloutGroupListGrid.java]: else is not necessary

File: hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/rollout/rolloutgroup/RolloutGroupListGrid.java [hunk 31]
```diff
@@ -0,0 +1,392 @@
...
+    };
+
+    private void onClickOfRolloutGroupName(RendererClickEvent event) {
+        rolloutUIState
+                .setRolloutGroup(rolloutGroupManagement.findRolloutGroupWithDetailedStatus((Long) event.getItemId()));
+        eventBus.publish(this, RolloutEvent.SHOW_ROLLOUT_GROUP_TARGETS);
+    }
+
+    private String convertRolloutGroupStatusToString(final RolloutGroupStatus value) {
+        String result = null;
```
Comment (SirWayne) [inline hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/rollout/rolloutgroup/RolloutGroupListGrid.java]: remove local paramter

File: hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/rollout/rolloutgrouptargets/RolloutGroupTargetsListGrid.java [hunk 32]
```diff
@@ -0,0 +1,309 @@
...
+
+import com.vaadin.data.Container;
+import com.vaadin.data.util.converter.Converter;
+import com.vaadin.server.FontAwesome;
+import com.vaadin.spring.annotation.SpringComponent;
+import com.vaadin.spring.annotation.ViewScope;
+
+@SpringComponent
+@ViewScope
+public class RolloutGroupTargetsListGrid extends AbstractGrid {
```
Comment (SirWayne) [inline hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/rollout/rolloutgrouptargets/RolloutGroupTargetsListGrid.java]: add javadoc

File: hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/rollout/rolloutgrouptargets/RolloutGroupTargetsListGrid.java [hunk 33]
```diff
@@ -0,0 +1,309 @@
...
+    }
+
+    @PreDestroy
+    void destroy() {
+        eventBus.unsubscribe(this);
+    }
+
+    @EventBusListenerMethod(scope = EventScope.SESSION)
+    void onEvent(final RolloutEvent event) {
+        if (event == RolloutEvent.SHOW_ROLLOUT_GROUP_TARGETS) {
```
Comment (SirWayne) [inline hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/rollout/rolloutgrouptargets/RolloutGroupTargetsListGrid.java]: 1. early return
2. constant for first paramter

File: hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/rollout/rolloutgrouptargets/RolloutGroupTargetsListGrid.java [hunk 34]
```diff
@@ -0,0 +1,309 @@
...
+    }
+
+    @Override
+    protected void addColumnRenderes() {
+        getColumn(SPUILabelDefinitions.VAR_STATUS).setRenderer(new HtmlLabelRenderer(), new StatusConverter());
+    }
+
+    @Override
+    protected void setHiddenColumns() {
+        /**
```
Comment (SirWayne) [inline hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/rollout/rolloutgrouptargets/RolloutGroupTargetsListGrid.java]: use normal comment not javadoc

File: hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/rollout/rolloutgrouptargets/RolloutGroupTargetsListGrid.java [hunk 35]
```diff
@@ -0,0 +1,309 @@
...
+         * No hidden columns.
+         */
+    }
+
+    @Override
+    protected CellDescriptionGenerator getDescriptionGenerator() {
+        return cell -> getDescription(cell);
+    };
```
Comment (SirWayne) [inline hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/rollout/rolloutgrouptargets/RolloutGroupTargetsListGrid.java]: remove ;

File: hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/rollout/rolloutgrouptargets/RolloutGroupTargetsListGrid.java [hunk 36]
```diff
@@ -0,0 +1,309 @@
...
+        public String convertToPresentation(final Status status, final Class<? extends String> targetType,
+                final Locale locale) {
+            if (status == null) {
+                // Actions are not created for targets when
+                // rollout's status
+                // is READY and when duplicate assignment is done.
+                // In these cases display a appropriate status with
+                // description
+                return getStatus();
+            } else {
```
Comment (SirWayne) [inline hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/rollout/rolloutgrouptargets/RolloutGroupTargetsListGrid.java]: else is not necessary

File: hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/rollout/rolloutgrouptargets/RolloutGroupTargetsListGrid.java [hunk 37]
```diff
@@ -0,0 +1,309 @@
...
+
+        @Override
+        public Class<String> getPresentationType() {
+            return String.class;
+        }
+
+    }
+
+    private String processActionStatus(final Status status) {
+        String result = null;
```
Comment (SirWayne) [inline hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/rollout/rolloutgrouptargets/RolloutGroupTargetsListGrid.java]: remove local variable

File: hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/rollout/rolloutgrouptargets/RolloutGroupTargetsListGrid.java [hunk 38]
```diff
@@ -0,0 +1,309 @@
...
+            return HawkbitCommonUtil.getStatusLabelDetailsInString(
+                    Integer.toString(FontAwesome.MINUS_CIRCLE.getCodepoint()), "statusIconBlue", null);
+        } else {
+            return HawkbitCommonUtil.getStatusLabelDetailsInString(
+                    Integer.toString(FontAwesome.QUESTION_CIRCLE.getCodepoint()), "statusIconBlue", null);
+        }
+    }
+
+    private String getDescription(CellReference cell) {
+        if (SPUILabelDefinitions.VAR_STATUS.equals(cell.getPropertyId())) {
```
Comment (SirWayne) [inline hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/rollout/rolloutgrouptargets/RolloutGroupTargetsListGrid.java]: early return

File: hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/utils/HawkbitCommonUtil.java [hunk 39]
```diff
@@ -1348,4 +1351,48 @@ private static boolean isNoTargets(final Long... statusCount) {
...
+     * Get the formatted string of status and target count details.
+     * 
+     * @param details
+     *            details of status and count
+     * @return String
+     */
+    public static String getFormattedString(Map<Status, Long> details) {
+        StringBuilder val = new StringBuilder();
+        String finalVal = null;
+        if (null != details && !details.isEmpty()) {
```
Comment (SirWayne) [inline hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/utils/HawkbitCommonUtil.java]: ealry return no local variable needed

File: hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/customrenderers/renderers/HtmlButtonRenderer.java [hunk 40]
```diff
@@ -0,0 +1,28 @@
...
+
+/**
+ * 
+ * Renders button with provided HTML content. Used to display button with icons.
+ *
+ */
+public class HtmlButtonRenderer extends ButtonRenderer {
+    private static final long serialVersionUID = -1242995370043404892L;
+
+    public HtmlButtonRenderer() {
```
Comment (SirWayne) [inline hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/customrenderers/renderers/HtmlButtonRenderer.java]: add javadoc

File: hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/customrenderers/renderers/HtmlButtonRenderer.java [hunk 41]
```diff
@@ -0,0 +1,28 @@
...
+        super();
+    }
+
+    public HtmlButtonRenderer(RendererClickListener listener) {
```
Comment (SirWayne) [inline hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/customrenderers/renderers/HtmlButtonRenderer.java]: add javadoc

File: hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/customrenderers/renderers/LinkRenderer.java [hunk 42]
```diff
@@ -0,0 +1,27 @@
...
+import com.vaadin.ui.renderers.ButtonRenderer;
+
+/**
+ * 
+ * Renders link with provided text.
+ *
+ */
+public class LinkRenderer extends ButtonRenderer {
+    private static final long serialVersionUID = -1242995370043404892L;
+    public LinkRenderer() {
```
Comment (SirWayne) [inline hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/customrenderers/renderers/LinkRenderer.java]: add javadoc

File: hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/customrenderers/renderers/LinkRenderer.java [hunk 43]
```diff
@@ -0,0 +1,27 @@
...
+        super();
+    }
+    public LinkRenderer(RendererClickListener listener) {
```
Comment (SirWayne) [inline hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/customrenderers/renderers/LinkRenderer.java]: add javadoc

File: hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/utils/HawkbitCommonUtil.java [hunk 44]
```diff
@@ -1348,4 +1351,47 @@ private static boolean isNoTargets(final Long... statusCount) {
...
+    /**
+     * Get the formatted string of status and target count details.
+     * 
+     * @param details
+     *            details of status and count
+     * @return String
+     */
+    public static String getFormattedString(Map<Status, Long> details) {
+        StringBuilder val = new StringBuilder();
+        if (null != details && !details.isEmpty()) {
```
Comment (SirWayne) [inline hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/utils/HawkbitCommonUtil.java]: if (details == null || details.isEmpty()) --> return

**Comment by SirWayne** [comment]
+1

**Comment by kaizimmerm** [comment]
@asharani-murugesh please fix the conflicts, thx.

**Pull request merged**